### PR TITLE
[118] add role to a freak

### PR DESCRIPTION
**Pull request by gabrielgiroe1**
## Description
 In this PR we can add a freak with a role

## Related

[[42] [Add Role to a Freak](https://trello.com/c/1NIOAZss/118-add-role-to-a-freak)

## Steps to Test or Reproduce
Go in insomnia and create a Freak.You can add this mutation:
```
mutation {
	freakCreate(
		firstName: "john",
		lastName: "doe",
		description: "ceva",
		email: "freak@gmail.com",
		roleId: 1
	) {
		firstName
		lastName
		description
		email
		role {
			name
		}
	}
}
```
Query freaks to see that the new freak was create:
```
query Freaks($last: Int, $before: String) {
    freaks(last: $last, before: $before) {
    
        edges {
            node {
                name
                firstName
                lastName
                description
                role {
                    name
                }       
            }
        }
    }
}
```

### CODE REVIEW

File: spec/controllers/graphql/queries/freaks_spec.rb [hunk 1]
```diff
@@ -13,29 +13,14 @@ module Graphql
     end
 
     before do
-      create(:freak, first_name: 'Ion')
-      create(:freak, first_name: 'Gheorghe')
+      role = create(:role, name: 'Software developer')
+      create(:freak, role_id: role.id)
```
Comment (alex-damian-negru) [inline spec/controllers/graphql/queries/freaks_spec.rb:17]: You can use FactoryBot traits to improve this.

But looking over the freak factory, it seems to have an association role, so is it really necessary to create a new role here since the freak already has one? o.o

**Review by alex-damian-negru — CHANGES_REQUESTED** [review]

**Review by salcudeann — APPROVED** [review]

File: db/seeds.rb [hunk 2]
```diff
@@ -7,3 +7,6 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+['Founder', 'IT Sibiu', 'IT Cluj', 'Team Assistant'].each do |role_name|
```
Comment (alex-damian-negru) [inline db/seeds.rb:10]: We might want to extract these values to a constant.

File: lib/tasks/add_freak_with_a_role.rake [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+namespace :add_freak_with_a_role do
```
Comment (alex-damian-negru) [inline lib/tasks/add_freak_with_a_role.rake:3]: Why is this needed? We can use the mutation to do the same thing.

File: spec/controllers/graphql/queries/freaks_spec.rb [hunk 4]
```diff
@@ -13,11 +13,16 @@ module Graphql
     end
 
     before do
-      create(:freak, first_name: 'Ion')
-      create(:freak, first_name: 'Gheorghe')
+      create(:freak)
     end
 
-    it { is_expected.to match_response_for(query: :freaks, sample: :freaks) }
+    context 'with freak role' do
```
Comment (alex-damian-negru) [inline spec/controllers/graphql/queries/freaks_spec.rb:19]: Considering the context, I'd expect outside the `freak` outside to have no role and inside of it to have a `role`, but I don't see any role being created inside the context & assigned to the `freak`.

Either make that clear for the context or move the test outside the context.

**Review by alex-damian-negru — DISMISSED** [review]

File: app/graphql/resolvers/create_freak_resolver.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Resolvers
+  class CreateFreakResolver < Resolvers::Base
```
Comment (alex-damian-negru) [inline app/graphql/resolvers/create_freak_resolver.rb:4]: I assume this is not used.

File: db/migrate/20210812120729_create_add_roles.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  def up
+    create_table :roles do |t|
+      t.string :name
+
+      t.timestamps
+    end
+    add_column :freaks, :role_id, :bigint
+    add_foreign_key :freaks, :roles
+
+    # add_index :roles, :id
```
Comment (alex-damian-negru) [inline db/migrate/20210812120729_create_add_roles.rb:13]: You forgot about this line.

File: db/migrate/20210813064244_rename_add_role_to_role.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class RenameAddRoleToRole < ActiveRecord::Migration[6.1]
```
Comment (alex-damian-negru) [inline db/migrate/20210813064244_rename_add_role_to_role.rb:3]: This migration does nothing.

File: lib/tasks/add_freak_with_a_role.rake [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+  desc 'Add a freak with a role'
+  task add_freak_in_db: :environment do
```
Comment (alex-damian-negru) [inline lib/tasks/add_freak_with_a_role.rake:5]: This does not properly describe what the method does. Make it clear, otherwise it will induce other devs in error.

File: spec/controllers/graphql/mutations/freak_create_spec.rb [hunk 9]
```diff
@@ -12,6 +12,12 @@ module Graphql
       }
     end
 
-    it { is_expected.to match_response_for(mutation: :freak_create, sample: :freakCreate) }
+    context 'when role exists' do
```
Comment (alex-damian-negru) [inline spec/controllers/graphql/mutations/freak_create_spec.rb:15]: This context would make one think about `when the roles does not exist` as well. Is that a valid case?

If not, I doubt this context makes sense to exist.

File: spec/controllers/graphql/queries/freaks_spec.rb [hunk 10]
```diff
@@ -13,11 +13,17 @@ module Graphql
     end
 
     before do
-      create(:freak, first_name: 'Ion')
-      create(:freak, first_name: 'Gheorghe')
+      create(:role, id: 1, name: 'Founder')
+      create(:freak, role_id: 1)
```
Comment (alex-damian-negru) [inline spec/controllers/graphql/queries/freaks_spec.rb:17]: A freak already has an associated role in the factory, so there's no need to re-create a new role.

File: spec/controllers/graphql/queries/freaks_spec.rb [hunk 11]
```diff
@@ -13,11 +13,17 @@ module Graphql
...
     end
 
-    it { is_expected.to match_response_for(query: :freaks, sample: :freaks) }
```
Comment (alex-damian-negru) [inline spec/controllers/graphql/queries/freaks_spec.rb:20]: Why delete this test? The behavior of the code did not change. The response still contains a role.

File: spec/fixtures/requests/queries/freaks.graphql [hunk 12]
```diff
@@ -13,10 +13,7 @@ query Freaks($last: Int, $before: String) {
                 lastName
                 description
                 email
-                photo {
-                    id
-                    uri
-                }
```
Comment (alex-damian-negru) [inline spec/fixtures/requests/queries/freaks.graphql:19]: Why delete this?

File: spec/fixtures/responses/queries/freaks/freaks_pagination.json [hunk 13]
```diff
@@ -12,15 +12,11 @@
...
-              "firstName": "Ion",
+              "name": "john doe",
+              "firstName": "john",
               "lastName": "doe",
               "description": "a simple freak",
               "email": "freak@gmail.com",
-              "photo": {
-                "id": "3",
-                "uri": "http://www.something.ro"
-              },
```
Comment (alex-damian-negru) [inline spec/fixtures/responses/queries/freaks/freaks_pagination.json:23]: Why delete this?

**Review by alex-damian-negru — CHANGES_REQUESTED** [review]

File: spec/fixtures/responses/mutations/freak_create/freak_create.json [hunk 14]
```diff
@@ -1,16 +1,11 @@
...
-        "lastName": "doe",
-        "description": "ceva",
-        "email": "freak@gmail.com",
-        "norm": {
-          "name": "full_time"
-        }
+    "data": null,
+    "errors": [
+      {
+        "message": "Cannot return null for non-nullable field Freak.role"
```
Comment (alex-damian-negru) [inline spec/fixtures/responses/mutations/freak_create/freak_create.json:7]: That's not what we expect. We expect the data from the `Freak` to be returned.

File: db/migrate/20210813064244_rename_add_role_to_role.rb (hunk 7 above)
Comment (alex-damian-negru) [reply db/migrate/20210813064244_rename_add_role_to_role.rb:3]: Don't forget to remove this.

**Review by alex-damian-negru — CHANGES_REQUESTED** [review]

File: db/migrate/20210813064244_rename_add_role_to_role.rb (hunk 7 above)
Comment (gabrielgiroe1) [reply db/migrate/20210813064244_rename_add_role_to_role.rb:3]: I will delete it

File: spec/fixtures/responses/mutations/freak_create/freak_create.json (hunk 14 above)
Comment (gabrielgiroe1) [reply spec/fixtures/responses/mutations/freak_create/freak_create.json]: Ok, I fix this

**Review by pirvudoru — APPROVED** [review]
